### PR TITLE
Add micropip + pyodide-py recipe

### PR DESCRIPTION
**Pull request by yuvipanda**
Generated by grayskull. Useful to try to use micropip to install packages from pypi when using https://github.com/jupyterlite/xeus-python-kernel and mamba / emscriptenforge

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by wolfv** [comment]
I think these lines might be an issue https://github.com/pyodide/micropip/blob/8c164d2bc4d3e140339257fcbe8a0d7e500d6521/micropip/_compat_in_pyodide.py#L5-L8 but we can think abotu that later.

CC @DerThorsten

**Comment by yuvipanda** [comment]
@wolfv hmm, the errors make me think this wasn't as unvendored from pyodide as I had thought. Which makes sense maybe.

File: recipes/micropip/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  noarch: python
+  script: {{ PYTHON }} -m pip install . -vv
+  number: 0
+
+requirements:
+  host:
+    - python >=3.7
+    - pip
+  run:
+    - python >=3.7
```
Comment (rth) [inline recipes/micropip/meta.yaml:22]: Thanks for doing this @yuvipanda!

Could you please add the following runtime requirements,
https://github.com/pyodide/pyodide/blob/main/packages/micropip/meta.yaml#L16

Looks like there were missing https://github.com/pyodide/micropip/issues/21 but we also didn't have many users of this package outside of Pyodide so far :)

Comment (rth) [reply recipes/micropip/meta.yaml:22]: And also https://pypi.org/project/pyodide-py/ is another dependency, you would likely need to package it as well (it's just another pure Python package).

Finally, Python 3.7+ is misleading, it actually needs 3.10+ . We could support a broader range of versions if there is demand for it.

**Comment by yuvipanda** [comment]
@rth thank you very much! I've added `pyodide-py` as a package in the same PR (lmk if it should be split into a different one), and set min supported version to `3.10`

**Comment by rth** [comment]
> And also [pypi.org/project/pyodide-py](https://pypi.org/project/pyodide-py/) is another dependency

This is only used [here](https://github.com/pyodide/micropip/blob/a2fad52e9ae063f4aed1e48213548f76988e24f5/micropip/_compat_in_pyodide.py#L10) and emscipten forge could have some runtime detection to use the available JS/Py FFI and not rely on pyodide. Still to make the CI pass, for now, probably the easiest is to package pyodide-py. 

It should not have any dependencies aside from the Python 3.10+  (and its stdlib).

**Comment by rth** [comment]
Thanks! LGTM. Let's see if CI passes.

**Comment by yuvipanda** [comment]
@rth looks like it passed, yay! Do you have rights to merge? :D

**Comment by yuvipanda** [comment]
@rth @wolfv uh, I should probably add other people as maintainers here as well. Can I add both of you? :D

**Comment by westurner** [comment]
- [ ] BLD,TST: micropip: A `micropip check` command
- [ ] BLD,TST: meta.yml: add `micropip check` test command

**Comment by westurner** [comment]
Is it justified to bundle pytest?

**Comment by yuvipanda** [comment]
@westurner the way I see it, it's super early in micropip's life, and getting this here would enable testing having both conda and pip packages mixed together in jupyterlite (which is currently not possible). I am not particularly involved in micropip, just a random user trying to get noarch packages from conda-forge and pure wheels from pypi working :) 

My hope is that this can get merged, and issues with health may be worked on one at a time. 

I also don't see they have pytest bundled in their source (https://github.com/pyodide/micropip/tree/main/micropip/). I also don't understand the comment in https://github.com/conda-forge/staged-recipes/pull/20961#issuecomment-1299522331. Can you explain?

**Comment by westurner** [comment]
> Is it justified to bundle pytest?

"Packages built in Pyodide" 
https://pyodide.org/en/stable/usage/packages-in-pyodide.html :

> pytest  7.1.2
> pytest-benchmark  

https://beta.mamba.pm/channels/emscripten-forge?tab=packages&size=25&index=2 :

> pytest	n/a	7.1.1
> pytest_driver	n/a	0.5.1
> pytest_driver_node	n/a	0.5.1

https://github.com/emscripten-forge/recipes#adding-packages

File: recipes/micropip/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    # From https://github.com/pyodide/pyodide/blob/main/packages/micropip/meta.yaml#L16
+    - packaging
+    - pyparsing
+    - pyodide-py
+
+test:
+  imports:
+    - micropip
+  commands:
+    - pip check
```
Comment (westurner) [inline recipes/micropip/meta.yaml:32]: Is there a `micropip check` command?

From https://github.com/conda-forge/staged-recipes/pull/20961#issuecomment-1299522331 :

> * [ ] BLD,TST: micropip: A `micropip check` command
>   
>   * https://pypi.org/project/pyodide-micropip-test/
> * [ ]  BLD,TST: meta.yml: add `micropip check` test command

Comment (yuvipanda) [reply recipes/micropip/meta.yaml:32]: @westurner I don't think there is.

Comment (westurner) [reply recipes/micropip/meta.yaml:32]: @wolfv what would `micropip check` check?.

**Comment by yuvipanda** [comment]
@westurner The `pyodide-py` package does not bundle pytest. I just verified this by installing `pyodide-py` in a blank venv, and the only files it installs are:

```
./lib/python3.10/site-packages/pyodide_py-0.22.0a1.dist-info
./lib/python3.10/site-packages/pyodide_py-0.22.0a1.dist-info/RECORD
./lib/python3.10/site-packages/pyodide_py-0.22.0a1.dist-info/WHEEL
./lib/python3.10/site-packages/pyodide_py-0.22.0a1.dist-info/top_level.txt
./lib/python3.10/site-packages/pyodide_py-0.22.0a1.dist-info/REQUESTED
./lib/python3.10/site-packages/pyodide_py-0.22.0a1.dist-info/INSTALLER
./lib/python3.10/site-packages/pyodide_py-0.22.0a1.dist-info/METADATA
./lib/python3.10/site-packages/_pyodide
./lib/python3.10/site-packages/_pyodide/_base.py
./lib/python3.10/site-packages/_pyodide/_core_docs.py
./lib/python3.10/site-packages/_pyodide/_importhook.py
./lib/python3.10/site-packages/_pyodide/__init__.py
./lib/python3.10/site-packages/_pyodide/__pycache__
./lib/python3.10/site-packages/_pyodide/__pycache__/_base.cpython-310.pyc
./lib/python3.10/site-packages/_pyodide/__pycache__/_core_docs.cpython-310.pyc
./lib/python3.10/site-packages/_pyodide/__pycache__/docstring.cpython-310.pyc
./lib/python3.10/site-packages/_pyodide/__pycache__/_importhook.cpython-310.pyc
./lib/python3.10/site-packages/_pyodide/__pycache__/__init__.cpython-310.pyc
./lib/python3.10/site-packages/_pyodide/docstring.py
./lib/python3.10/site-packages/pyodide
./lib/python3.10/site-packages/pyodide/_package_loader.py
./lib/python3.10/site-packages/pyodide/console.py
./lib/python3.10/site-packages/pyodide/_run_js.py
./lib/python3.10/site-packages/pyodide/code.py
./lib/python3.10/site-packages/pyodide/__init__.py
./lib/python3.10/site-packages/pyodide/__pycache__
./lib/python3.10/site-packages/pyodide/__pycache__/_core.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/__pycache__/_state.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/__pycache__/_run_js.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/__pycache__/_package_loader.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/__pycache__/code.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/__pycache__/webloop.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/__pycache__/console.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/__pycache__/http.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/__pycache__/__init__.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/_core.py
./lib/python3.10/site-packages/pyodide/_state.py
./lib/python3.10/site-packages/pyodide/ffi
./lib/python3.10/site-packages/pyodide/ffi/__init__.py
./lib/python3.10/site-packages/pyodide/ffi/__pycache__
./lib/python3.10/site-packages/pyodide/ffi/__pycache__/wrappers.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/ffi/__pycache__/__init__.cpython-310.pyc
./lib/python3.10/site-packages/pyodide/ffi/wrappers.py
./lib/python3.10/site-packages/pyodide/http.py
./lib/python3.10/site-packages/pyodide/webloop.py
```

`pyodide-py` isn't all of pyodide, just a bunch of shims

**Comment by yuvipanda** [comment]
@westurner https://github.com/conda-forge/staged-recipes/pull/20961#discussion_r1010855167 has more information on what the pyodide-py package is.

**Comment by rth** [comment]
No, I don't have merge permissions here. Yes, please add me as a co-maintainer.

**Comment by ocefpaf** [comment]
Dumb question: What is the value of micropip outside of the browser? I thought it was only need inside a pyodide distribution.

**Comment by rth** [comment]
> What is the value of micropip outside of the browser?

Not much, this is mostly so that emsctipen-forge could use it via the noarch channel as far as I understand.

The pyodide-py is more relevant, as people can `import pyodide` in notebooks and then those notebooks would fail to work outside of the browser without this package being installed. Jupyterlite tries to smooth this issue by obfuscating most of this, but it still can be an issue, which is the initial reason we uploaded it to PyPI

**Comment by ocefpaf** [comment]
Thanks @rth!

**Comment by ocefpaf** [comment]
> Useful to try to use micropip to install packages from pypi when using https://github.com/jupyterlite/xeus-python-kernel and mamba / emscriptenforge

Argh. The problem with so many boilerplate PR/Issues. @yuvipanda had already answered my question and I missed it. Sorry Yuvi!

**Comment by rth** [comment]
Thanks @ocefpaf !

**Comment by yuvipanda** [comment]
THANK YOU SO MUCH, @ocefpaf and @rth!

**Pull request merged**